cpp/src/arrow/compute/kernels/scalar_set_lookup.cc

```diff
@@ -36,16 +36,23 @@ namespace {
 
 template <typename Type>
 struct SetLookupState : public KernelState {
-  explicit SetLookupState(MemoryPool* pool) : lookup_table(pool, 0) {}
+  explicit SetLookupState(MemoryPool* pool) : lookup_table(pool, 0), memory_pool(pool) {}
```

Comment:
Ok, but this makes an initial memory allocation for the lookup table which is immediately undone in `Init`.

A possible way around that is to make `lookup_table` a `std::optional<MemoTable>`, to allow deferred initialization.

Reply:
Good point. Changed it to `std::optional`